homeassistant/components/anova_sous_vide/__init__.py

```diff
@@ -0,0 +1,40 @@
+"""The Anova integration."""
+from __future__ import annotations
+
+from anova_wifi import AnovaOffline, AnovaPrecisionCooker, AnovaPrecisionCookerSensor
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import DOMAIN
+from .coordinator import AnovaCoordinator
+
+PLATFORMS = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Anova from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+    apc = AnovaPrecisionCooker()
+    try:
+        update = await apc.update(entry.data["device_id"])
+    except AnovaOffline as ex:
+        raise ConfigEntryNotReady("Can not connect to the sous vide") from ex
```

Comment:
Create the coordinator, and than call 
`await coordinator.async_config_entry_first_refresh()`

It will do the same thing as this block
```suggestion
    apc = AnovaPrecisionCooker()
```

Reply:
How can I do this when I need the update for creating the coordinator? I need to get the firmware version.

Reply:
How can I do this when I need the update for creating the coordinator? I need to get the firmware version.

Reply:
Refactor the coordinator to set the device info after the first `async_config_entry_first_refresh`

Reply:
```diff
diff --git a/homeassistant/components/anova_sous_vide/__init__.py b/homeassistant/components/anova_sous_vide/__init__.py
index 8fa812789a..376946227a 100644
--- a/homeassistant/components/anova_sous_vide/__init__.py
+++ b/homeassistant/components/anova_sous_vide/__init__.py
@@ -1,12 +1,11 @@
 """The Anova integration."""
 from __future__ import annotations
 
-from anova_wifi import AnovaOffline, AnovaPrecisionCooker, AnovaPrecisionCookerSensor
+from anova_wifi import AnovaPrecisionCooker, AnovaPrecisionCookerSensor
 
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import Platform
 from homeassistant.core import HomeAssistant
-from homeassistant.exceptions import ConfigEntryNotReady
 
 from .const import DOMAIN
 from .coordinator import AnovaCoordinator
@@ -17,16 +16,12 @@ PLATFORMS = [Platform.SENSOR]
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Anova from a config entry."""
 
-    hass.data.setdefault(DOMAIN, {})
     apc = AnovaPrecisionCooker()
-    try:
-        update = await apc.update(entry.data["device_id"])
-    except AnovaOffline as ex:
-        raise ConfigEntryNotReady("Can not connect to the sous vide") from ex
-    hass.data[DOMAIN][entry.entry_id] = AnovaCoordinator(
-        hass,
-        AnovaPrecisionCooker(),
-        update["sensors"][AnovaPrecisionCookerSensor.FIRMWARE_VERSION],
+    coordinator = AnovaCoordinator(hass, apc)
+    await coordinator.async_config_entry_first_refresh()
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+    coordinator.async_setup(
+        coordinator.data["sensors"][AnovaPrecisionCookerSensor.FIRMWARE_VERSION],
     )
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
     return True
diff --git a/homeassistant/components/anova_sous_vide/coordinator.py b/homeassistant/components/anova_sous_vide/coordinator.py
index 32b397fd69..05de4b3063 100644
--- a/homeassistant/components/anova_sous_vide/coordinator.py
+++ b/homeassistant/components/anova_sous_vide/coordinator.py
@@ -5,7 +5,7 @@ import logging
 from anova_wifi import AnovaOffline, AnovaPrecisionCooker
 import async_timeout
 
-from homeassistant.core import HomeAssistant
+from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity import DeviceInfo
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
@@ -21,7 +21,6 @@ class AnovaCoordinator(DataUpdateCoordinator):
         self,
         hass: HomeAssistant,
         anova_api: AnovaPrecisionCooker,
-        firmware_version: str,
     ) -> None:
         """Set up Anova Coordinator."""
         super().__init__(
@@ -30,8 +29,14 @@ class AnovaCoordinator(DataUpdateCoordinator):
             logger=_LOGGER,
             update_interval=timedelta(seconds=30),
         )
-        if self.config_entry is not None:
+        assert self.config_entry is not None
         self._device_id = self.config_entry.data["device_id"]
+        self.anova_api = anova_api
+        self.device_info: DeviceInfo | None = None
+
+    @callback
+    def async_setup(self, firmware_version: str) -> None:
+        """Set the firmware version info."""
         self.device_info = DeviceInfo(
             identifiers={(DOMAIN, self._device_id)},
             name="Anova Precision Cooker",
@@ -39,10 +44,6 @@ class AnovaCoordinator(DataUpdateCoordinator):
             model="Precision Cooker",
             sw_version=firmware_version,
         )
-        else:
-            _LOGGER.error("Anova Coordinator was setup without config entry")
-
-        self.anova_api = anova_api
 
     async def _async_update_data(self):
         try:
```

Reply:
Thank you, this was super helpful!